Common/cpp/SharedItems/Shareables.h

```diff
@@ -216,96 +253,79 @@ class ShareableHostFunction : public Shareable {
+        function_(std::move(function)) {}
+
   jsi::Value toJSValue(jsi::Runtime &rt) override {
-    if (runtimeHelper_->isUIRuntime(rt)) {
+    if (&rt == runtime_) {
+      return jsi::Value(rt, function_);
+    } else {
 #ifdef DEBUG
-      return runtimeHelper_->valueUnpacker->call(
+      return getValueUnpacker(rt).call(
           rt,
           ShareableJSRef::newHostObject(rt, shared_from_this()),
           jsi::String::createFromAscii(rt, "RemoteFunction"));
 #else
       return ShareableJSRef::newHostObject(rt, shared_from_this());
 #endif
-    } else {
-      return jsi::Value(rt, function_);
     }
   }
 };
 
 class ShareableHandle : public Shareable {
  private:
-  std::shared_ptr<JSRuntimeHelper> runtimeHelper_;
   std::unique_ptr<ShareableObject> initializer_;
   std::unique_ptr<jsi::Value> remoteValue_;
+  jsi::Runtime *remoteRuntime_;
 
  public:
-  ShareableHandle(
-      const std::shared_ptr<JSRuntimeHelper> runtimeHelper,
-      jsi::Runtime &rt,
-      const jsi::Object &initializerObject)
-      : Shareable(HandleType), runtimeHelper_(runtimeHelper) {
-    initializer_ = std::make_unique<ShareableObject>(rt, initializerObject);
+  ShareableHandle(jsi::Runtime &rt, const jsi::Object &initializerObject)
+      : Shareable(HandleType),
+        initializer_(std::make_unique<ShareableObject>(rt, initializerObject)) {
   }
+
   ~ShareableHandle() {
-    if (runtimeHelper_->uiRuntimeDestroyed) {
-      // The below use of unique_ptr.release prevents the smart pointer from
-      // calling the destructor of the kept object. This effectively results in
-      // leaking some memory. We do this on purpose, as sometimes we would keep
-      // references to JSI objects past the lifetime of its runtime (e.g.,
-      // shared values references from the RN VM holds reference to JSI objects
-      // on the UI runtime). When the UI runtime is terminated, the orphaned JSI
-      // objects would crash the app when their destructors are called, because
-      // they call into a memory that's managed by the terminated runtime. We
-      // accept the tradeoff of leaking memory here, as it has a limited impact.
-      // This scenario can only occur when the React instance is torn down which
-      // happens in development mode during app reloads, or in production when
-      // the app is being shut down gracefully by the system. An alternative
-      // solution would require us to keep track of all JSI values that are in
-      // use which would require additional data structure and compute spent on
-      // bookkeeping that only for the sake of destroying the values in time
-      // before the runtime is terminated. Note that the underlying memory that
-      // jsi::Value refers to is managed by the VM and gets freed along with the
-      // runtime.
+    if (remoteRuntime_ != nullptr &&
+        !WorkletRuntimeRegistry::isRuntimeAlive(remoteRuntime_)) {
+      // See comment in `~RetainingShareable()`.
```

Comment:
Maybe we can just have a method `cleanupIfRuntimeExists`, call it from here, from`~RetainingShareable` and other places (`~ShareableSynchronizedDataHolder` etc). Then we could have the comment placed in one place only?

Reply:
Great idea, done in 6ce4b7b02943f7b0310f8dac014a064d541ac8c3